File: revconnect/social-service/src/main/java/com/revconnect/socialservice/service/FollowService.java
package com.revconnect.socialservice.service;

import com.revconnect.socialservice.dto.FollowDTO;
import com.revconnect.socialservice.entity.Follow;
import com.revconnect.socialservice.repository.FollowRepository;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class FollowService {

    private final FollowRepository followRepository;

    public FollowService(FollowRepository followRepository) {
        this.followRepository = followRepository;
    }

    // Creates a PENDING follow request
    @Transactional
    public FollowDTO followUser(Long followerId, Long followingId) {
        if (followerId.equals(followingId)) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Cannot follow yourself");
        }
        if (followRepository.findByFollowerIdAndFollowingId(followerId, followingId).isPresent()) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "Follow request already sent or already following");
        }
        Follow follow = new Follow();
        follow.setFollowerId(followerId);
        follow.setFollowingId(followingId);
        follow.setStatus("PENDING");
        return convertToDTO(followRepository.save(follow));
    }

    // Accept a pending follow request
    @Transactional
    public FollowDTO acceptFollow(Long followId) {
        Follow follow = followRepository.findById(followId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Follow request not found"));
        if (!"PENDING".equals(follow.getStatus())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Request is not pending");
        }
        follow.setStatus("ACCEPTED");
        return convertToDTO(followRepository.save(follow));
    }

    // Reject/delete a pending follow request
    @Transactional
    public void rejectFollow(Long followId) {
        Follow follow = followRepository.findById(followId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Follow request not found"));
        followRepository.delete(follow);
    }

    @Transactional
    public void unfollowUser(Long followerId, Long followingId) {
        Follow follow = followRepository.findByFollowerIdAndFollowingId(followerId, followingId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Follow relationship not found"));
        followRepository.delete(follow);
    }

    // Returns only ACCEPTED followers
    public List<Long> getFollowers(Long userId) {
        return followRepository.findByFollowingIdAndStatus(userId, "ACCEPTED").stream()
                .map(Follow::getFollowerId)
                .collect(Collectors.toList());
    }

    // Returns only ACCEPTED following
    public List<Long> getFollowing(Long userId) {
        return followRepository.findByFollowerIdAndStatus(userId, "ACCEPTED").stream()
                .map(Follow::getFollowingId)
                .collect(Collectors.toList());
    }

    // Pending follow requests received by userId
    public List<FollowDTO> getPendingFollowRequests(Long userId) {
        return followRepository.findByFollowingIdAndStatus(userId, "PENDING").stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    // Check if followerId is following followingId (ACCEPTED only)
    public boolean isFollowing(Long followerId, Long followingId) {
        return followRepository.findByFollowerIdAndFollowingId(followerId, followingId)
                .map(f -> "ACCEPTED".equals(f.getStatus()))
                .orElse(false);
    }

    // Check raw follow status (PENDING or ACCEPTED)
    public String getFollowStatus(Long followerId, Long followingId) {
        return followRepository.findByFollowerIdAndFollowingId(followerId, followingId)
                .map(Follow::getStatus)
                .orElse("NONE");
    }

    private FollowDTO convertToDTO(Follow follow) {
        FollowDTO dto = new FollowDTO();
        dto.setId(follow.getId());
        dto.setFollowerId(follow.getFollowerId());
        dto.setFollowingId(follow.getFollowingId());
        dto.setStatus(follow.getStatus());
        dto.setCreatedAt(follow.getCreatedAt());
        return dto;
    }
}

File: revconnect-frontend/src/app/services/follow.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, tap } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from '../../environments/environment';
import { AuthService } from './auth.service';

export interface Follow {
  id: number;
  followerId: number;
  followingId: number;
  status: string;
  createdAt: string;
}

@Injectable({ providedIn: 'root' })
export class FollowService {
  private apiUrl = `${environment.apiUrl}/follows`;

  constructor(private http: HttpClient, private authService: AuthService) {}

  // Sends a PENDING follow request
  followUser(followerId: number, followingId: number): Observable<Follow> {
    return this.http.post<Follow>(this.apiUrl, { followingId }).pipe(
      tap(() => {
        const username = this.authService.getCurrentUser()?.username || 'Someone';
        this.http.post(`${environment.apiUrl}/notifications`, {
          senderId: followerId,
          receiverId: followingId,
          type: 'FOLLOW',
          message: `${username} sent you a follow request`,
          postId: null
        }).subscribe({ error: () => {} });
      })
    );
  }

  unfollowUser(followerId: number, followingId: number): Observable<void> {
    return this.http.delete<void>(this.apiUrl, { params: { followingId } });
  }

  acceptFollow(followId: number): Observable<Follow> {
    return this.http.put<Follow>(`${this.apiUrl}/${followId}/accept`, {});
  }

  rejectFollow(followId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${followId}/reject`);
  }

  getPendingFollowRequests(): Observable<Follow[]> {
    return this.http.get<Follow[]>(`${this.apiUrl}/pending/received`);
  }

  getPendingFollowCount(): Observable<number> {
    return this.getPendingFollowRequests().pipe(map(r => r.length));
  }

  getFollowers(userId: number): Observable<number[]> {
    return this.http.get<number[]>(`${this.apiUrl}/followers/${userId}`);
  }

  getFollowing(userId: number): Observable<number[]> {
    return this.http.get<number[]>(`${this.apiUrl}/following/${userId}`);
  }

  checkFollowStatus(followerId: number, followingId: number): Observable<{ isFollowing: boolean }> {
    return this.http.get<boolean>(`${this.apiUrl}/check`, { params: { followerId, followingId } }).pipe(
      map(result => ({ isFollowing: result }))
    );
  }

  // Returns NONE | PENDING | ACCEPTED
  getFollowStatus(followerId: number, followingId: number): Observable<string> {
    return this.http.get<{ status: string }>(`${this.apiUrl}/status`, { params: { followerId, followingId } }).pipe(
      map(r => r.status)
    );
  }
}
